exec/I18N.hs

```diff
@@ -0,0 +1,393 @@
+        f old =
+            case List.filter ((== newimport) . snd) old of
+                (_:_) -> old
+                [] -> case List.filter (ST.isPrefixOf "import " . snd) old of
+                    ((i, _):_) -> case splitAt i old of
+                        (before, after) -> before <> [(-1, newimport), (-1, "")] <> after
+
+    moduleNameFromFilePath :: FilePath -> ST
+    moduleNameFromFilePath = cs . (<> "_T") . dropExtension . intercalate "." . tail . splitDirectories
+
+quoteString :: ST -> ST
+quoteString = ("\"" <>) . (<> "\"") . ST.replace "\"" "\\\""
+
+
+-- * main
+
+translationsFile :: FilePath
+translationsFile = "./aula-i18n.json"  -- TODO: make this configurable?
+
+main :: IO ()
+main = do
+    wd <- getEnv "AULA_ROOT_PATH"  -- TODO: get rid of this, too.
+    args <- getArgs
+    run wd args
+
+run :: FilePath -> [String] -> IO ()
+run wd = \case
+    []                 -> refresh Nothing
+    [diffFrom, diffTo] -> refresh (Just (cs diffFrom, cs diffTo))
+  where
+    refresh mdiff = do
+        putStrLn $ "looking for existing " <> show translationsFile <> "..."
+        table :: TransTableF <- do
+            e <- doesFileExist translationsFile
+            s <- if e then either (pure mempty) id . Aeson.eitherDecode <$> LBS.readFile translationsFile
+                      else pure mempty
+            s `seq` pure s
+
+        putStrLn $ "looking for new translation keys" <> maybe "" (\_ -> " and translations") mdiff <> "..."
+        table' <- (<> table) <$> extractTransTableF whereToGetTheLangValue wd mdiff
+
+        putStrLn $ "updating " <> show translationsFile <> "..."
+        LBS.writeFile translationsFile (Aeson.encodePretty table')
+
+        putStrLn "generating `*_T` modules..."
+        compileTransTableF table'
+
+        putStrLn "done!"
+
+
+-- TODO: make this a separate package thentos-i18n
+
+t1 :: IO ()
+t1 = do
+    wd <- getEnv "AULA_ROOT_PATH"
+    run wd ["9a5c4b5b~1", "9a5c4b5b"]
+
+
+conv :: Aeson.ToJSON a => (Lang -> TransTableF -> a) -> IO ()
+conv mk = LBS.readFile translationsFile
+      >>= \t -> forM_ [minBound..] $ \lang ->
+                  LBS.writeFile ("aula-transifex-" <> show lang <> ".json") (mk' lang t)
+  where
+    mk' lang = Aeson.encodePretty . mk lang . fromMaybe (error "conv") . Aeson.decode
+
+data Transifex = Transifex Lang TransTableF
+  deriving (Show, Eq)
+
+instance Aeson.ToJSON Transifex where
+  toJSON (Transifex lang f) = gotf f
+    where
+      gotf (TransTableF m) = Aeson.object . fmap (uncurry (Aeson..=) . (_1 %~ cs)) . Map.toList . Map.map gotk $ m
+
+      gotk :: TransTableK -> Aeson.Value
+      gotk (TransTableK m) = Aeson.object $ uncurry gott <$> Map.toList m
+
+      gott :: TKey -> TransTableT -> Aeson.Pair
+      gott k (TransTableT (ls, m)) = ak Aeson..= av
+        where
+          ak = k <> ":" <> ST.intercalate ":" (cs . show <$> Set.toAscList ls)
+          av = fromMaybe (error "Transifex: incomplete translation table") $ Map.lookup lang m
```

Comment:
The `lang` parameter should be included in the message to give the user some context.